modules/local/hipeak/prepare_counts.nf

```diff
@@ -0,0 +1,185 @@
+    #######################################################################
+    #######################################################################
+    ## Created on Aug. 24, 2021 count reads for peak filtering
+    ## Copyright (c) 2021 Jianhong Ou (jianhong.ou@gmail.com)
+    #######################################################################
+    #######################################################################
+    pkgs <- c("rtracklayer", "InteractionSet", "Biostrings", "Rsamtools")
+    versions <- c("${getProcessName(task.process)}:")
+    for(pkg in pkgs){
+        # load library
+        library(pkg, character.only=TRUE)
+        # parepare for versions.yml
+        versions <- c(versions,
+            paste0("    ", pkg, ": ", as.character(packageVersion(pkg))))
+    }
+    writeLines(versions, "versions.yml") # wirte versions.yml
+
+    ## options
+    ## make_option(c("-a", "--r1peak"), type="character", default=NULL, help="filename of r1 peak", metavar="string")
+    ## make_option(c("-b", "--r2peak"), type="character", default=NULL, help="filename of r2 peak", metavar="string")
+    ## make_option(c("-x", "--restrict"), type="character", default=NULL, help="filename of restrict cut", metavar="string")
+    ## make_option(c("-p", "--pairs"), type="character", default=NULL, help="folder of valid distal pairs", metavar="string")
+    ## make_option(c("-m", "--mappability"), type="character", default=NULL, help="mappability file", metavar="string")
+    ## make_option(c("-o", "--output"), type="character", default="counts.csv", help="output folder", metavar="string")
+    ## make_option(c("-f", "--fasta"), type="character", default=NULL, help="genome fasta file", metavar="string")
+    OUTPUT <- "counts.${meta.id}.csv"
+    FASTA <- "$fasta"
+    CUT <- "$cut"
+    MAPPABILITY <- "$mappability"
+    pattern <- "unselected.pairs.gz" ## unselected.pairs.gz is postfix of output of pairtools filter
+    pairs <- dir("pairs", paste0(pattern, "\$"), full.names=TRUE)
+    names(pairs) <- sub(paste0("\\\\.", pattern), "", basename(pairs))
+    R1PEAK <- import("$r1peak")
+    R2PEAK <- import("$r2peak")
+    mcols(R1PEAK) <- NULL
+    mcols(R2PEAK) <- NULL
+    ## split by chromsome
+    R1PEAK <- split(R1PEAK, seqnames(R1PEAK))
+    R2PEAK <- split(R2PEAK, seqnames(R2PEAK))
+    R1PEAK <- R1PEAK[lengths(R1PEAK)>0]
+    R2PEAK <- R2PEAK[lengths(R2PEAK)>0]
+    chromosomes <- intersect(names(R1PEAK), names(R2PEAK))
+    chromosomes <- chromosomes[!grepl("_", chromosomes)]
+    chromosomes <- chromosomes[!grepl("M", chromosomes)] ## remove chrM/chrMT
+    if(length(chromosomes)==0){
+        stop("no valid data in same chromosome.")
+    }
+
+    ## loading data
+    readPairs <- function(pair, chrom1, chrom2){
+        fn <- paste(pair, chrom1, chrom2, "rds", sep=".")
+        if(file.exists(paste0(pair, ".rds"))){
+            if(file.exists(fn)){
+                pc <- readRDS(fn)
+            }else{
+                NULL
```

Comment:
This code is used to speed up the IO process. To avoid to load the whole huge file into memory during whole process, I first split the files by chromosomes and then load the files when required.

Reply:
Sorry, I specifically meant the `NULL`. It's not assigned to anything, or do I miss where it is assigned?

Reply:
Oh, sorry for the short script. That means it will return a NULL. I will fix this by assign it to variable to make it readable.